src/lib/sql/sql_translator.cpp

```diff
@@ -403,15 +410,103 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_join(const hsql::Join
                                                    : std::make_pair(*left_in_right_node, *right_in_left_node);
 
   // Joins currently only support one simple condition (i.e., not multiple conditions).
-  auto predicate_condition = translate_operator_type_to_predicate_condition(condition.opType);
-
+  auto predicate_condition = translate_operator_type_to_predicate_condition(join_condition->opType);
   auto join_node = JoinNode::make(join_mode, column_references, predicate_condition);
   join_node->set_left_input(left_node);
   join_node->set_right_input(right_node);
-
+  _apply_predicate_expressions(join_node, LQPInputSide::Left, left);
+  _apply_predicate_expressions(join_node, LQPInputSide::Right, right);
   return join_node;
 }
 
+LQPInputSide SQLTranslator::_get_side(const std::shared_ptr<AbstractLQPNode>& left_node,
+                                      const std::shared_ptr<AbstractLQPNode>& right_node, const hsql::Expr* expr) {
+  DebugAssert(expr && expr->type == hsql::kExprColumnRef, "only applicable to columnar expressions");
+  const auto qualified = HSQLExprTranslator::to_qualified_column_name(*expr);
+  if (left_node->find_column(qualified)) {
+    return LQPInputSide::Left;
+  }
+
+  if (right_node->find_column(qualified)) {
+    return LQPInputSide::Right;
+  }
+
+  Fail("Column not found: " + qualified.as_string());
+}
+
+void SQLTranslator::_get_sides_from_operator_expressions(std::vector<const hsql::Expr*> operators,
+                                                         std::function<LQPInputSide(const hsql::Expr*)> get_side,
+                                                         std::vector<const hsql::Expr*>& both,
+                                                         std::vector<const hsql::Expr*>& left,
+                                                         std::vector<const hsql::Expr*>& right) {
+  for (const auto* op : operators) {
+    const bool left_is_column = op->expr && op->expr->type == hsql::kExprColumnRef;
+    const bool right_is_column = op->expr2 && op->expr2->type == hsql::kExprColumnRef;
+
+    if (left_is_column && right_is_column) {
+      auto left_column_side = get_side(op->expr);
+      auto right_column_side = get_side(op->expr2);
+      if (left_column_side == right_column_side) {
+        if (left_column_side == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else {
+        both.push_back(op);
+      }
+    } else {
+      if (left_is_column) {
+        if (get_side(op->expr) == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else if (right_is_column) {
```

Comment:
Wouldn't this mean that the predicate `1 = 2` would fall through and not get added to the list at all, falsifying the result?

Reply:
Yepp, it's broken :/

```
(debug)> SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND t1.a = 12
=== Columns
|       a|       b|       a|       b|
|     int|   float|     int|   float|
=== Chunk 0 ===
|      12|   350.7|      12|   350.7|
===
1 rows total (COMPILE: 254 µs, EXECUTE: 1987 µs (wall time))
(debug)> SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND 11 = 12
=== Columns
|       a|       b|       a|       b|
|     int|   float|     int|   float|
=== Chunk 0 ===
|      12|   350.7|      12|   350.7|
=== Chunk 1 ===
|     123|   458.7|     123|   458.7|
=== Chunk 2 ===
|   12345|   456.7|   12345|   456.7|
|   12345|   457.7|   12345|   456.7|
|   12345|   456.7|   12345|   457.7|
|   12345|   457.7|   12345|   457.7|
===
6 rows total (COMPILE: 143 µs, EXECUTE: 1857 µs (wall time))
```

Please add a test that checks for this.

Reply:
OK I shouldn't have skipped the expression in the first place. FYI `_translate_predicate` enforces that at least one of the operands is a column.